module/zfs/brt.c

```diff
@@ -0,0 +1,1705 @@
+ * each free we have to check if there is a corresponding entry in the BRT or
+ * not. If there is, we need to decrease the reference counter. Doing BRT
+ * lookup on every free can potentially be expensive by requiring additional
+ * I/Os if the BRT doesn't fit into memory. This is the main problem with
+ * deduplication, so we've learn our lesson and try not to repeat the same
+ * mistake here. How do we do that? We divide each top-level VDEV into 1GB
+ * regions. For each region we maintain a reference counter that is a sum of
+ * all reference counters of the cloned blocks that have offsets within the
+ * region. This creates the regions array of 64bit numbers for each top-level
+ * VDEV. The regions array is always kept in memory and updated on disk in the
+ * same transaction group as the BRT updates to keep everything in-sync. We can
+ * keep the array in memory, because it is very small. With 1GB regions and 1TB
+ * VDEV the array requires only 8kB of memory (we may decide to decrease the
+ * region size in the future). Now, when we want to free a block, we first
+ * consult the array. If the counter for the whole region is zero, there is no
+ * need to look for the BRT entry, as there isn't one for sure. If the counter
+ * for the region is greater than zero, only then we will do a BRT lookup and
+ * if an entry is found we will decrease the reference counters in the entry
+ * and in the regions array.
+ *
+ * The regions array is small, but can potentially be larger for very large
+ * VDEVs or smaller regions. In this case we don't want to rewrite entire array
+ * on every change. We then divide the regions array into 128kB chunks and keep
+ * a bitmap of dirty chunks within a transaction group. When we sync the
+ * transaction group we can only update the parts of the regions array that
+ * were modified. Note: Keeping track of the dirty parts of the regions array
+ * is implemented, but updating only parts of the regions array on disk is not
+ * yet implemented - for now we will update entire regions array if there was
+ * any change.
+ *
+ * The implementation tries to be economic: if BRT is not used, or no longer
+ * used, there will be no entries in the MOS and no additional memory used (eg.
+ * the regions array is only allocated if needed).
+ *
+ * Interaction between Deduplication and Block Cloning.
+ *
+ * If both functionalities are in use, we could end up with a block that is
+ * referenced multiple times in both DDT and BRT. When we free one of the
+ * references we couldn't tell where it belongs, so we would have to decide
+ * what table takes the precedence: do we first clear DDT references or BRT
+ * references? To avoid this dilemma BRT cooperates with DDT - if a given block
+ * is being cloned using BRT and the BP has the D (dedup) bit set, BRT will
+ * lookup DDT entry and increase the counter there. No BRT entry will be
+ * created for a block that resides on a dataset with deduplication turned on.
```

Comment:
I'll try to clarify that. You are right that we may have a block residing on dedup-enabled dataset and not having the D bit set (ie. it was created before dedup was turned on).
Also note that it is impossible to set the D bit after block creation, so "once a block has been added to the BRT it can't be added to the DDT" cannot happen as there is no way to add a block to DDT after it was created.